File: cpp/leetcode/318.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int n = words.size();
        int res = 0;
        vector<int> masks(words.size(),0);
        for(int i = 0; i < n; i++)
        	for(char ch : words[i]){
        		masks[i] |= 1 << (ch - 'a');
        	}

        for(int i = 0; i < n; i++)
        	for(int j = i+1; j < n; j++){
        		if (!(masks[i] & masks[j])){
        			res = max(res, int(words[i].size() * words[j].size()));
        		}
        	}
        return res;
    }
};

File: python/leetcode/318.py
class Solution:
    def maxProduct(self, words):
        n, res = len(words), 0
        masks = [0] * n
        for i in xrange(n):
            for ch in words[i]:
                masks[i] |= 1 << (ord(ch) - ord('a'))
        # each masks[i] has 32 digits which is enough to save 26 characters
        # so masks[i] records each alphabeta characters in words[i]
        for i in xrange(n):
            for j in xrange(i + 1, n):
                if (masks[i] & masks[j]) == 0:
                    res = max(res, len(words[i]) * len(words[j]))
        return res